KAFKA-19932: adding handling of OOM and avoiding wrapped as timeout

iit2009060 opened the pull request:
adding handling of OOM and avoiding wrapped as timeout

iit2009060 committed:
adding handling of OOM and avoiding wrapped as timeout

iit2009060 commented:
@kirktrue can you review it

kirktrue reviewed (commented):
Thanks for the PR @iit2009060. Can you add some unit tests to validate the work? Thanks!

iit2009060 committed:
adding test case for OOM detection

iit2009060 commented:
> Thanks for the PR @iit2009060. Can you add some unit tests to validate the work? Thanks!

added

iit2009060 commented:
@kirktrue  can you review it.

iit2009060 commented:
@kirktrue  can you please review it

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -11749,4 +11750,84 @@ public void testDescribeTopicsTimeoutWhenNoBrokerResponds() throws Exception {
+            // The AdminClient calls response.topicMetadata() in listTopics handleResponse(), which will trigger the OOM
+            MetadataResponseData data = new MetadataResponseData();
+            MetadataResponse realResponse = new MetadataResponse(data, ApiKeys.METADATA.latestVersion());
+            MetadataResponse spyResponse = spy(realResponse);
```
Is it possible to use a generic "mock" here instead of a spy?

iit2009060 replied:
done

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -11749,4 +11750,84 @@ public void testDescribeTopicsTimeoutWhenNoBrokerResponds() throws Exception {
+            // We expect ExecutionException wrapping the OutOfMemoryError (this is standard Future behavior)
+            ExecutionException exception = assertThrows(ExecutionException.class, () -> result.names().get());
+            assertInstanceOf(OutOfMemoryError.class, exception.getCause(),
+                    "Expected OutOfMemoryError to be propagated, but got: " + exception.getCause());
```
Can you see if you can use TestUtils.assertFutureThrows() here? It could be slightly less boilerplate.

iit2009060 replied:
done

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -11749,4 +11750,84 @@ public void testDescribeTopicsTimeoutWhenNoBrokerResponds() throws Exception {
+            ExecutionException exception = assertThrows(ExecutionException.class, () -> result.names().get());
+            assertInstanceOf(OutOfMemoryError.class, exception.getCause(),
+                    "Expected OutOfMemoryError to be propagated, but got: " + exception.getCause());
+            assertEquals("Simulated OOM during response handling", exception.getCause().getMessage());
```
Another option is to create the `OutOfMemoryError` outside of the `doThrow()` and then perform `assertEquals()`.

kirktrue reviewed (commented):
@iit2009060 Sorry for the delay. Thanks for the addition of the unit tests.

I'm going back and forth on whether we should handle this outside of `handleTimeoutFailure()` or not. On the one hand, the root cause (OOM) is not a timeout, but on the other hand, we might consider hitting an OOM to be a "retriable" error. Thoughts?

Also, I provided some minor feedback on the tests.

Thanks!

kirktrue commented:
@iit2009060—gentle reminder on this. Thanks!

kirktrue commented:
@iit2009060—gentle reminder on this. Thanks!

iit2009060 commented:
> @iit2009060—gentle reminder on this. Thanks!

@kirktrue  Thanks for reminding me. I will have a look.

kirktrue commented:
@iit2009060—do you still have the time and motivation to work on this, or should we hand it off to someone else? I just want to make sure we get it into Kafka 4.3. Thanks!

iit2009060 committed:
Converting spy to mock

iit2009060 commented:
> @iit2009060—do you still have the time and motivation to work on this, or should we hand it off to someone else? I just want to make sure we get it into Kafka 4.3. Thanks!

i am really sorry if you have to follow up on this multiple times.I have addressed the comments.

iit2009060 commented:
> I'm going back and forth on whether we should handle this outside of `handleTimeoutFailure()` or not. On the one hand, the root cause (OOM) is not a timeout, but on the other hand, we might consider hitting an OOM to be a "retriable" error. Thoughts?
Yes i think OOM is not a retryable error and we should not be handling in a timeout  failure.  We should fail fast instead of waiting on the timeout. Should I make a change for this ?

kirktrue commented:
> Yes i think OOM is not a retryable error and we should not be handling in a timeout failure. We should fail fast instead of waiting on the timeout. Should I make a change for this ?

Yes, please.

Nikita-Shupletsov commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -956,7 +955,10 @@ private void handleTimeoutFailure(long now, Throwable cause) {
             }
-            if (cause instanceof TimeoutException) {
+            // Don't mask OutOfMemoryError as TimeoutException - propagate it directly
+            if (cause instanceof OutOfMemoryError) {
```
is there any reason to wrap any errors into a TImeoutException?
I may be wrong, but it looks like it may make sense to treat all errors the same way.
wdyt?

iit2009060 replied:
@Nikita-Shupletsov  I am moving Out of memory error out of Timeout  as out of memory is not a retryable error.

Nikita-Shupletsov replied:
yeah, I understand that, sorry for not being crisp. My question was more like: as we found that problem with OutOfMemoryError, should we treat all errors the same way? I am not sure if StackOverFlowError for example is a retriable error in that context, or NotSuchMethodError.

iit2009060 replied:
Yes that should also not be treated a timedout error , we can open a separate JIRA to handle this specific errors to move out of  timedout exception.

kirktrue replied:
@iit2009060—Can you create that separate PR?

kirktrue replied:
IIUC, @Nikita-Shupletsov is suggesting that we catch `VirtualMachineError`.

iit2009060 replied:
@kirktrue  can we track in a separate ticket? This PR has been open for a long time.

Nikita-Shupletsov replied:
@iit2009060 
May I ask what your concern is? it should be a one-liner. I am not 100% sure why we would need a separate ticket for that

iit2009060 replied:
@Nikita-Shupletsov  Are you suggesting instead of catching OutOfMemoryError , we should catch VirtualMachineError which avoid all other non retryable exception to fall under the timedout error? 

<img width="893" height="205" alt="Screenshot 2026-03-11 at 10 50 39 PM" src="https://github.com/user-attachments/assets/6184ffd8-5f6e-4d0a-b10f-b86a2b9e640e" />

Nikita-Shupletsov replied:
effectively yes

iit2009060 replied:
@Nikita-Shupletsov  @kirktrue, I have made the changes. Please check

iit2009060 committed:
Update OOM handling outside of timeout

iit2009060 commented:
> > Yes i think OOM is not a retryable error and we should not be handling in a timeout failure. We should fail fast instead of waiting on the timeout. Should I make a change for this ?
> 
> Yes, please.

done

iit2009060 commented:
@kirktrue  Please review it , I have made the changes.

iit2009060 committed:
Catch VirtualMachineError instead of OOM

iit2009060 commented:
@Nikita-Shupletsov  @kirktrue  i have made the changes ,Please review it.

Nikita-Shupletsov reviewed (approved):
LGTM. thank you

iit2009060 commented:
> LGTM. thank you

@kirktrue  Can you also review it so we can make it to 4.3

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -11787,4 +11787,68 @@ public void testDescribeTopicsTimeoutWhenNoBrokerResponds() throws Exception {
+
+            // Now advance time to exceed the timeout - this tests that even when timeout would occur,
+            // the OOM error that was already processed takes precedence over TimeoutException
+            time.sleep(1500);
```
Is there a reason that the admin client has to wait for the timeout to get the OOM error?

iit2009060 replied:
No i think this timeout does not makes sense , as OOM is already thrown. I am removing this redundant test.

kirktrue reviewed (commented):
@iit2009060—looks good to me. Just one question about the timeout behavior in `testOutOfMemoryErrorNotMaskedOnTimeout()`.

iit2009060 committed:
removing redundant test

kirktrue reviewed (approved):
Thanks for your patience @iit2009060!
